plans/caching.go

```diff
@@ -0,0 +1,158 @@
+package plans
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
+	"reflect"
+	"strings"
+
+	"github.com/memcachier/mc"
+	"github.com/outdoorsy/gorp"
+)
+
+const defaultCacheExpirationTime = 0 // never expire
+
+func prepareForCache(data interface{}) (string, error) {
+	var b bytes.Buffer
+	w := gzip.NewWriter(&b)
+	err := json.NewEncoder(w).Encode(encodeForMemcache(data))
+	if err != nil {
+		w.Close()
+		return "", err
+	}
+	err = w.Flush()
+	if err != nil {
+		w.Close()
+		return "", err
+	}
+	err = w.Close()
+	if err != nil {
+		return "", err
+	}
+	return string(b.Bytes()), nil
+}
+
+func restoreFromCache(encoded string, target reflect.Value, table *gorp.TableMap) ([]interface{}, error) {
+	r, err := gzip.NewReader(strings.NewReader(encoded))
+	if err != nil {
+		r.Close()
+		return nil, err
+	}
+
+	var data interface{}
+	err = json.NewDecoder(r).Decode(&data)
+	if err != nil {
+		return nil, err
+	}
+	err = r.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	var elem reflect.Value
+	if target.Kind() == reflect.Ptr {
+		elem = reflect.Zero(target.Type())
+	} else {
+		elem = reflect.New(target.Type().Elem())
+	}
```

Comment:
Actually, the conditions here should be reversed - if it's a pointer, you want to use `reflect.New`; if it's not a pointer, you want to use `reflect.Zero`.  I probably should have caught that earlier.